src/v1/atm/atm-update-service.js

```diff
@@ -0,0 +1,64 @@
+import request from 'request';
+import Atm from './atm-model';
+import atmDao from './atm-dao';
+import logger from '../../lib/logger';
+
+const allAtmUrl =
+  ' https://api.privatbank.ua/p24api/infrastructure?json&atm&address=&city=';
+
+class PrivatbankAtmService {
+  constructor(url, DAO) {
+    this.url = url;
+    this.DAO = DAO;
+  }
+
+  async getAllAtmsFromPrivatbankApi() {
+    logger.info('Do request to Privatbank for ATMs data.');
+    const promise = new Promise((resolve, reject) => {
+      request.get(this.url, (error, responce, body) => {
+        if (error) {
+          reject(error);
+        }
+        const result = JSON.parse(body).devices;
+        logger.info('Data of ATMs Privatbank were received.');
+        resolve(result);
+      });
+    });
+    return promise;
+  }
+
+  async updatePartOfAtmsDataInDB(partOfAtmsData) {
+    const promises = [];
+    partOfAtmsData.forEach((atmData) => {
+      const atm = new Atm();
+      atm.setData(atmData);
+      promises.push(this.DAO.create(atm));
+    });
+    await Promise.all(promises);
+  }
+
+  async distributedAdditionInDB(atmsData, number) {
+    const STEP = 1000;
+    logger.info(
+      `Do update ATM [${number};${Math.min(
+        number + STEP,
+        atmsData.length
+      )}] from ${atmsData.length} ATMs`
+    );
+
+    await this.updatePartOfAtmsDataInDB(atmsData.slice(number, number + STEP));
+    if (atmsData.length > number + STEP) {
+      await this.distributedAdditionInDB(atmsData, number + STEP);
+    }
+  }
+
+  async updateAtmsDataInDB() {
+    const atmsData = await this.getAllAtmsFromPrivatbankApi();
+    await this.DAO.deleteAll(); // TODO: Do behavior without dropping db on few seconds
```

Comment:
@v1adko , here, before insert new data I delete all from ATM collection, I think, that this is not correct and i should update(or insert new) all existing data and then remove all what were outdated. But this way a little harder and I have some problem with it. For example atms doesn't have unique fields, and it's do finding harder.
What do you say? Should I change this logic or not?

Reply:
Yes, that's wrong.

You can treat object's [hash](https://www.npmjs.com/package/object-hash) as an identifier and store it.
So before update, map over the data from PB api, calculate and assign hashes, then load all ATM hashes from our DB (using $select and $lean) to speed it up and assign transform that into a [Map](https://developer.mozilla.org/uk/docs/Web/JavaScript/Reference/Global_Objects/Map) and then filter the new documents by looking up their hashes in the map (map lookup is fast, O(1)), resulting in only new objects.

BTW this function will be executed very few times coz the ATM db is not updated that frequently.
There are possible hash collisions, but those are unlikely and having some duplicate hash documents is not that big of a problem that can be checked and fixed quite easily after the fact.

Reply:
Exactly! hash! Big thank you, I forgot about this way. :)